src/bundles/ar/AR.ts

```diff
@@ -0,0 +1,272 @@
+}
+
+/**
+ * Obtains AR module state.
+ *
+ * @returns Current AR state.
+ */
+export function getModuleState(): ARState {
+  return (window as any).arController as ARState;
+}
+
+/**
+ * Calls callback to update AR context in tab.
+ */
+export function callARCallback() {
+  const f = (window as any).arControllerCallback as Function;
+  if (f) {
+    f();
+  }
+}
+
+// Overlay
+
+/**
+ * Sets the left toggle.
+ *
+ * @param text Label on toggle.
+ * @param callback Function to call when toggle is clicked.
+ */
+export function setLeftToggle(text: string, callback: () => void) {
+  const moduleState = getModuleState();
+  if (!moduleState) return;
+  moduleState.overlay.toggleLeft = new Toggle(text, callback);
+}
+
+/**
+ * Sets the center toggle.
+ *
+ * @param text Label on toggle.
+ * @param callback Function to call when toggle is clicked.
+ */
+export function setCenterToggle(text: string, callback: () => void) {
+  const moduleState = getModuleState();
+  if (!moduleState) return;
+  moduleState.overlay.toggleCenter = new Toggle(text, callback);
+}
+
+/**
+ * Sets the right toggle.
+ *
+ * @param text Label on toggle.
+ * @param callback Function to call when toggle is clicked.
+ */
+export function setRightToggle(text: string, callback: () => void) {
+  const moduleState = getModuleState();
+  if (!moduleState) return;
+  moduleState.overlay.toggleRight = new Toggle(text, callback);
+}
+
+/**
+ * Resets and hides the left toggle.
+ */
+export function removeLeftToggle() {
+  const moduleState = getModuleState();
+  if (!moduleState) return;
+  moduleState.overlay.toggleLeft = undefined;
+}
+
+/**
+ * Resets and hides the center toggle.
+ */
+export function removeCenterToggle() {
+  const moduleState = getModuleState();
+  if (!moduleState) return;
+  moduleState.overlay.toggleCenter = undefined;
+}
+
+/**
+ * Resets and hides the right toggle.
+ */
+export function removeRightToggle() {
+  const moduleState = getModuleState();
+  if (!moduleState) return;
+  moduleState.overlay.toggleRight = undefined;
+}
+
+// Objects
+
+/**
+ * Creates an instance of Vector3.
+ *
+ * @param x Value of x-axis.
+ * @param y Value of y-axis.
+ * @param z Value of z-axis.
+ * @returns Vector3 created from specified values.
+ */
+export function createVector3(x: number, y: number, z: number): Vector3 {
+  return new Vector3(x, y, z);
+}
+
+/**
+ * Adds the specified object to the augmented world.
+ *
+ * @param object ARObject to add. (E.g. cube, sphere, etc..)
+ */
+export function addARObject(object: ARObject) {
+  const moduleState = getModuleState();
+  if (!moduleState) return;
+  if (moduleState.arObjects.find((item) => item.id === object.id)) {
+    return; // Already in array
+  }
+  if (object.onSelect) {
+    moduleState.clickCallbacks.set(object.id, () => {
+      object.onSelect?.(object);
+    });
+  }
+  const newArray = Object.assign([], moduleState.arObjects);
+  newArray.push(object);
+  moduleState.arObjects = newArray;
```

Comment:
Does the `moduleState` need to be immutable? If there is no specific reason to keep it immutable under the hood, then we could actually make it mutable to improve performance (Tried rebuilding the module using mutable moduleState, works fine for me). 

Similar for other instances in AR.ts where the moduleState.arObjects is completely replaced, like `removeARObject()`

Reply:
The arObjects here can be modified while the AR context is running. (For example, one could add new objects to the world at the click of a button)